### ISSUE TITLE
Repository rules aren't refetched when their repo mappings change

### ISSUE THREAD

**Issue by fmeum** [bazelbuild/bazel#20722]
### Description of the bug:

This is similar to https://github.com/bazelbuild/bazel/issues/20721, but 1) about repository rules rather than module extensions and 2) not a regression in 7.0.0.

A repository rule isn't refetched when its repo mapping changes, which can result in `Label(...)` calls made in the implementation function to return stale results when canonical repo names change, e.g. during `bazel_dep` version bumps.

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```starlark
# MODULE.bazel
bazel_dep(name = "rules_go", version = "0.40.0")

ext = use_extension("//:ext.bzl", "ext")
ext.tag(label_str = "@rules_go//go")
use_repo(ext, "repo")

# BUILD
load("@repo//:defs.bzl", "LABEL")
print(LABEL)

# ext.bzl
def _repo_impl(ctx):
    ctx.file("WORKSPACE")
    ctx.file("BUILD")
    ctx.file("defs.bzl", "LABEL = Label({})".format(repr(str(Label(ctx.attr.label_str)))))

_repo = repository_rule(
    implementation=_repo_impl,
    attrs={"label_str": attr.string(mandatory=True)},
)

_tag = tag_class(
    attrs={"label_str": attr.string(mandatory=True)},
)

def _ext_impl(ctx):
    for module in ctx.modules:
        for tag in module.tags.tag:
            _repo(name = "repo", label_str = tag.label_str)

ext = module_extension(
    implementation=_ext_impl,
    tag_classes={"tag": _tag},
)
```

```shell
$ bazel build //... --lockfile_mode=off
DEBUG: /home/fhenneke/tmp/repo-mapping-lock/BUILD:2:6: @@rules_go~0.40.0//go:go
$ buildozer 'set version 0.41.0' MODULE.bazel:rules_go
$ bazel build //... --lockfile_mode=off
<no output since the old warning has already been omitted and hasn't changed>
$ bazel shutdown
$ bazel build //... --lockfile_mode=off
DEBUG: /home/fhenneke/tmp/repo-mapping-lock/BUILD:2:6: @@rules_go~0.40.0//go:go
```

### What is the output of `bazel info release`?

7.0.0 and HEAD

**Comment by fmeum** [comment]
@bazel-io fork 7.0.1

**Comment by fmeum** [comment]
If you don't plan to solve this for 7.0.1, we could alternatively fix the more immediate issue by getting rid of the `rules_java` loads in all Bazel-provided repo rules, instead relying on the native symbols.

**Comment by fmeum** [comment]
Feel free to downgrade this to 7.1.0 if you think that https://github.com/bazelbuild/bazel/pull/20810 is reasonable in the meantime.

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
